### ISSUE TITLE
Add alternative ranking method, and don't crash rendering ranking plot.

### ISSUE THREAD

**Issue by jonathanmetzman** [google/fuzzbench#219]
Didn't get a chance to investigate but generating a report for either experiment alone works:
Repro command:
```
GOOGLE_CLOUD_PROJECT=fuzzbench PYTHONPATH=. python analysis/generate_report.py -q -d /tmp/output -q 2020-04-08-aflplusplus 2020-04-01
```
I'm guessing it's because wpantund benchmark wasn't used in 2020-04-08-aflplusplus (maybe that should be an exception, but it should be clearer what's the reason).

**Comment by jonathanmetzman** [comment]
> I'm guessing it's because wpantund benchmark wasn't used in 2020-04-08-aflplusplus (maybe that should be an exception, but it should be clearer what's the reason).

Nope, this also fails:

```
GOOGLE_CLOUD_PROJECT=fuzzbench PYTHONPATH=. python analysis/generate_report.py -q -b bloaty_fuzz_target -d /tmp/2020-04-08-aflplusplus-and-2020-04-01 -q 2020-04-01 2020-04-08-aflplusplus
```

**Comment by jonathanmetzman** [comment]
To narrow it done some, it's worth noting this command succeeds:
```
GOOGLE_CLOUD_PROJECT=fuzzbench PYTHONPATH=. python analysis/generate_report.py -q -f aflplusplus aflplusplus_mopt aflplusplus_noalloc aflplusplus_fast aflplusplus_rare aflplusplus_lin aflplusplus_quad aflplusplus_coe aflplusplus_cmplog -b bloaty_fuzz_target curl_curl_fuzzer_http freetype2-2017 harfbuzz-1.3.2 irssi_server-fuzz jsoncpp_jsoncpp_fuzzer lcms-2017-03-21 libjpeg-turbo-07-2017 libpcap_fuzz_both libpng-1.2.56 libxml2-v2.9.2 mbedtls_fuzz_dtlsclient openssl_x509 openthread-2019-12-23 php_php-fuzz-parser proj4-2017-08-14 re2-2014-12-09 sqlite3_ossfuzz systemd_fuzz-link-parser vorbis-2017-12-11 woff2-2016-05-06 zlib_zlib_uncompress_fuzzer -d /tmp/2020-04-08-aflplusplus-and-2020-04-01 2020-04-01 2020-04-08-aflplusplus
```

**Comment by vanhauser-thc** [comment]
@jonathanmetzman  - quick question different to the issue but about the report.

the report shows that aflplusplus_mopt did run (among others), but I do not see aflplusplus_mmopt (note the the extra m). Is this correct? If yes, can you please run that one and add it? because either that or I have to setup all the fuzzers for another run for comparison. only the _mopt one (1 m :) ) is not needed.
the naming is a bit unfortunate I think ;)

**Comment by vanhauser-thc** [comment]
oh and the aflplusplus/ one seems to be a wrong old config, not the new one I pushed (unless I made a mistake)

**Comment by jonathanmetzman** [comment]
> @jonathanmetzman - quick question different to the issue but about the report.
> 
> the report shows that aflplusplus_mopt did run (among others), but I do not see aflplusplus_mmopt (note the the extra m). Is this correct? If yes, can you please run that one and add it? because either that or I have to setup all the fuzzers for another run for comparison. only the _mopt one (1 m :) ) is not needed.
> the naming is a bit unfortunate I think ;)

It did run, I simply forgot to include it. It's updated now (you might need to open the inspector to view new report since caching may serve you the old one).

>oh and the aflplusplus/ one seems to be a wrong old config, not the new one I pushed (unless I made a mistake)

It's the afl++ from [2020-04-01](https://www.fuzzbench.com/reports/2020-04-01/index.html) (which was run with this version of fuzzbench: https://github.com/google/fuzzbench/commits/1dd0281cab220ffa188308d76006f46288da1725). Is this older than you wanted?

**Comment by vanhauser-thc** [comment]
mmopt: thanks!

aflplusplus: yeah it had laf-intel, which I removed and the others dont have it either so we cannot compare it :)

**Comment by jonathanmetzman** [comment]
Sorry, give me 5 minutes with the report, i think the version that is up still doesn't have mmopt

**Comment by jonathanmetzman** [comment]
Actually the caching fooled me, opening in another browser shows the results with mmopt.
http://fuzzbench.com/reports/2020-04-08-aflplusplus-and-2020-04-01/index.html should have results with mmopt.

**Comment by jonathanmetzman** [comment]
I'll try to run another experiment with current afl++ and afl++_mopt and combine with these variants soon (I'm OOO again later this week so I'll probably have results by Friday or Monday).

**Comment by vanhauser-thc** [comment]
hmm we could also do that differently.

I could remove the fuzzers and add 3 x 4 variantes (mmopt, rare, explore schedule combined with default, instrim, ctx, ngram-2 and ngram-4 instrumentation).

that would be the next step anyway.
is that ok with you?

**Comment by andreafioraldi** [comment]
@vanhauser-thc IMO the experiments with schedules are enough. @jonathanmetzman apart from the graphs, I'm interested to see if different schedules discover unique paths that others cannot explore. IMO the defulat schedule (explore) is better when running a single instance like in fuzzbench, but as schedules are useful in parallel fuzzing I'm interesting in seeing how much the coverage overlap. Are these data available in fuzzbench? Can you give me a link to it?

**Comment by vanhauser-thc** [comment]
@andreafioraldi yes they are enough as the missing one is included in the next run I pushed in the night and just updated.
and that is a very good topic you are bringing up. there is a data, I think the bucket is this one:
https://console.cloud.google.com/storage/browser/fuzzbench-data/2020-04-08-aflplusplus

**Comment by vanhauser-thc** [comment]
@jonathanmetzman different topic - again ;) - how is the end/overall rating of the fuzzers performed? is it like the average of the placement of the fuzzer? if it is that I would propose something different.

problem: if a fuzzer is only in front by 1 edge over all others and in a few cases only has 1/2 of everyone else then the overall rating will still be good.

IMHO it would be better if the best fuzzer in a case would get 100 points, and all other fuzzers depending on the % of the edge count of the best fuzzer get their points. I think currently honggfuzz would benefit most from this but I still think that is a better rating.

**Comment by jonathanmetzman** [comment]
> @vanhauser-thc IMO the experiments with schedules are enough. @jonathanmetzman apart from the graphs, I'm interested to see if different schedules discover unique paths that others cannot explore. IMO the default schedule (explore) is better when running a single instance like in fuzzbench, but schedules are useful in parallel fuzzing and I'm interesting in seeing how much the coverage overlap. Are these data available in fuzzbench? Can you give me a link to it?

We stopped saving which edges were covered in each trial (but doing this again should be fairly straightforward).  But as @vanhauser-thc points out we do keep the corpora from each trial which you can find in the link he shared. Long-term we're looking at some metric for finding how unique the coverage of a particular fuzzer is (#77) but this is a bit of a low priority for now

**Comment by jonathanmetzman** [comment]
>@jonathanmetzman different topic - again ;) - how is the end/overall rating of the fuzzers performed? is it like the average of the placement of the fuzzer? if it is that I would propose something different.

Yes I believe the aggregate ranking is the average rank of a fuzzer on each benchmark (@lszekeres implemented this and should know more if needed).

>problem: if a fuzzer is only in front by 1 edge over all others and in a few cases only has 1/2 of everyone else then the overall rating will still be good.

>IMHO it would be better if the best fuzzer in a case would get 100 points, and all other fuzzers depending on the % of the edge count of the best fuzzer get their points. I think currently honggfuzz would benefit most from this change ;) but I still think that is a better rating.

Right I agree with the problem and I've had similar thoughts about a better aggregate ranking.
I tend to think the individual benchmarks are more important than the aggregate ranking for the reason you laid out though it's obviously not as convenient.

I think we are planning on adding more ways of interpreting the data to reports in the future (@lszekeres also has done a lot of work on this already) since there will never be a "perfect" metric. However, we only launched using one metric to keep things simple and understandable.
I do think that the ranking method you proposed would be better, but we will look at more metrics in the future.

**Comment by lszekeres** [comment]
> > @jonathanmetzman different topic - again ;) - how is the end/overall rating of the fuzzers performed? is it like the average of the placement of the fuzzer? if it is that I would propose something different.
> 
> Yes I believe the aggregate ranking is the average rank of a fuzzer on each benchmark (@lszekeres implemented this and should know more if needed).

Yep, there's also a more detailed description in the [docs](https://google.github.io/fuzzbench/reference/report/#experiment-summary
).

It's important to note that the ranking should only be interpreted by considering the statistical significance as well. I.e., if a fuzzer is not further from another than the "critical difference", then it shouldn't be considered "better". This is also the reason why we use average ranks: we rely on a statistical test that is commonly used for cross benchmark/dataset analysis in order to establish the significance of the results (namely the Friedman/Nemenyi test). Using average ranks is the most typical technique in non-parametric statistical tests - when we cannot assume normal distribution - such as the [Friedman test](https://en.wikipedia.org/wiki/Friedman_test
), the [Wilcoxon signed-rank test](https://en.wikipedia.org/wiki/Wilcoxon_signed-rank_test), or the [Kruskal-Wallis test](https://en.wikipedia.org/wiki/Kruskal%E2%80%93Wallis_one-way_analysis_of_variance). The issue you point out with this is a valid concern. It might worth noting though, that since we have a relatively high number of benchmarks, those small differences you mentioned should even out.

This is not to say that one ranking algorithm is better than the other. Our analysis library already includes some alternative ranking techniques (e.g., ranking by number of statistical tests wins), and we can add the one you proposed as well, using the normalized coverage numbers. The disadvantage of these alternatives is that we cannot apply the above mentioned "standard" statistical tests to establish the significance of the results. 

Having said that, I'm happy to add this ranking alternative too to the analysis library, if there's still interest, so folks can play with it.

**Comment by vanhauser-thc** [comment]
@lszekeres thank you for the detailed reply!
why you say adding that ranking to the analysis library - would that ranking be included in the report? Or is that ranking accessible via the cloud data bucket then?
For me the described ranking would be the one I would assess which feature to push and/or make default in afl++ 3.0

One other analysis that would be very interesting: number of edges a fuzzer found that no other fuzzer found (but that should ignore variants of the same fuzzer). e.g. number of edges only honggfuzz found or only entropic. not for ranking but additional information. I think fairfuzz would stand out there for example.

**Comment by jonathanmetzman** [comment]
> @lszekeres thank you for the detailed reply!
> why you say adding that ranking to the analysis library - would that ranking be included in the report? Or is that ranking accessible via the cloud data bucket then?

Laszlo would know better than I but I think there are plans to have a detailed report with many different rankings and metrics. This could go there, or might be the ranking we use in the main report.
 
For the record, there are also plans to include some analyses in our library that we wouldn't put in a report. Users can then run the analyses themselves on the csv file we link to in every report (Look for "You can download the raw data for this report here.", [example](https://www.fuzzbench.com/reports/2020-04-14/data.csv.gz). 

> For me the described ranking would be the one I would assess which feature to push and/or make default in afl++ 3.0

Good datapoint. btw, I think you said you don't do much Python, but obviously we would love it if we could get a patch for this ranking.

> 
> One other analysis that would be very interesting: number of edges a fuzzer found that no other fuzzer found (but that should ignore variants of the same fuzzer). e.g. number of edges only honggfuzz found or only entropic. not for ranking but additional information. I think fairfuzz would stand out there for example.

Yeah I think #77 tracks this feature.

**Comment by lszekeres** [comment]
> > @lszekeres thank you for the detailed reply!
> > why you say adding that ranking to the analysis library - would that ranking be included in the report? Or is that ranking accessible via the cloud data bucket then?
> 
> Laszlo would know better than I but I think there are plans to have a detailed report with many different rankings and metrics. This could go there, or might be the ranking we use in the main report.

Yes, I added the ranking alternative you suggested to the analysis library in https://github.com/google/fuzzbench/pull/236.

Currently, one easy way to use this is by making a copy of [report_templates/default.html](https://github.com/google/fuzzbench/blob/master/analysis/report_templates/default.html
), and adding [experiment.rank_by_median_and_average_normalized_score](https://github.com/google/fuzzbench/blob/master/analysis/experiment_results.py#L128) to the template. Then you can follow the [doc on how to generate](https://google.github.io/fuzzbench/developing-fuzzbench/custom_analysis_and_reports#generating-reports) your own reports from the raw data we link in all published reports. You can specify your template using the --report-type flag. I'm also planning to add an experimental report template that could include all these alternative analysis results (https://github.com/google/fuzzbench/issues/267), and also a Jupyter/Colab notebook for manual analyses (https://github.com/google/fuzzbench/issues/268).

**Comment by lszekeres** [comment]
> I'm guessing it's because wpantund benchmark wasn't used in 2020-04-08-aflplusplus (maybe that should be an exception, but it should be clearer what's the reason).

The cause of this crash was that the critical difference plot rendering implementation that we currently use only supports up to 20 fuzzers. This is not documented, the code just uses the internal statistical table which is looked up to compute the critical difference based on the number of fuzzers, and too many will trigger an OOB.
 
Fixed in https://github.com/google/fuzzbench/pull/272
